core/templates/dev/head/pages/exploration_editor/history_tab/HistoryServices.js

```diff
@@ -198,12 +198,23 @@ oppia.factory('compareVersionsService', [
           v1Data: $http.get(explorationDataUrl + v1),
           v2Data: $http.get(explorationDataUrl + v2)
         }).then(function(response) {
-          var v1States = response.v1Data.data.states;
-          var v2States = response.v2Data.data.states;
+          var v1StatesDict = response.v1Data.data.states;
+          var v2StatesDict = response.v2Data.data.states;
 
           // Track changes from v1 to LCA, and then from LCA to v2.
           var lca = versionsTreeService.findLCA(v1, v2);
 
+          var v1States = {};
+          var v2States = {};
+          for (var stateName in v1StatesDict) {
```

Comment:
Nope, I mean adding a new StatesObjectFactory file. Stuff in StateObjectFactory should pertain only to an individual state.

Thinking about it, might actually be worth doing in this PR, otherwise you'll have to find all the instances of this pattern again. Up to you, though -- either way is fine, but please file an issue if you don't address it in this PR so we don't forget about it.